### PR TITLE
improved energy saving by omitting p2p connections

### PR DESCRIPTION
**Pull request by PhilippEngler**
Hello @bropat,

with an energy saving device (esd) (in my case a T8170), I ran in the issue, that the device lost 10 to 20 percent of battery per day. Also for the esd, a p2p connection will be built up each 10 minutes. I did some investigation and have found the problem [here](https://github.com/bropat/eufy-security-client/blob/27e7c87d67810667022cf6cabb0c46ab1b818654/src/eufysecurity.ts#L359C13-L359C49). The `getStorageInfo` will build up a p2p connection, even if the device is not connected (like this is in the case of esd).

Because for the esd it is not necessary, I have changed the function, so that for these devices the p2p connection is omitted for a given value. If we use the standard value for `pollingIntervalMinutes` (10 minutes) and for (the new) `refreshP2PEnergySavingOmits` (2), the number of p2p connections will decrease from 6 to 2 per hour. The battery lost 2 to 6 percent per day.

Also I have implemented it in the way, if the esd is connected an this function is called, the counter is reset (the data has been updated, so the omits counter can start at 1). Anyway, by setting `refreshP2PEnergySavingOmits` to 0, the omits are disabled.

If you have any concerns or questions, please contact me.

Best, Philipp

### CODE REVIEW

**Comment by bropat** [comment]
I have corrected the implementation of the refresh via P2P. It is now carried out every 12 hours, as is already the case for other devices, unless a connection has already been established, in which case a refresh is carried out immediately and the 12-hour timer is reinitialised.
So if someone executes a P2P command, a P2P connection is established and all P2P data is immediately updated and the timer is reset.
Basically the previous `getStorageInfo` pull was corrected (#293). Its "refresh" implementation did not correspond to the correct refresh policy of the project.

**Comment by bropat** [comment]
@PhilippEngler 

Please try this implementation and let me know if it runs correctly (current data vs battery consumption).

**Comment by PhilippEngler** [comment]
Thank you, @bropat.

One issue I had right at the beginning. For two of my devices (T8416 and T8170), the storage data is retrieved at second run of the `refreshP2PData` function (For testing, I have reduced the interval to 2 minutes). During the first run after startup, the value for the property `StationSdStatus` is `undefined` (and so the `station.getStorageInfoEx()` will not executed).

After inserting `station.getStorageInfoEx();` to the `addStation`-function, the values are available after first run (but I'm not sure if this is a good workaround and if this would have implications to your ioBroker project).

Now I let it run for a day and then we will see...

**Comment by bropat** [comment]
@PhilippEngler 

> During the first run after startup, the value for the property StationSdStatus is undefined (and so the station.getStorageInfoEx() will not executed).

Fixed.

**Comment by PhilippEngler** [comment]
Thank you, @bropat.

I have no explanation why this does not work for me, but I had already removed the `undefinied`-check there and it did not work in my case. The T8170 (the only standalone energy saving device I have) got 0GB for total and used storage. So I have added the `getStorageInfoEx()` command to the `addStation`-function and it works in my case.

I apologize for the inconvenience.

**Comment by bropat** [comment]
@PhilippEngler 

Can you send me a debug log of T8170?

**Comment by PhilippEngler** [comment]
@bropat: send by email.

For now it seems ok, maybe it was a timing issue. I will keep an eye on.

**Comment by PhilippEngler** [comment]
@bropat: The storage info issue occurred today, you should have already received an email with the logs attached.
Thank you.

**Comment by bropat** [comment]
@PhilippEngler 

I have checked the logs and the problem seems to be an incorrect response at the beginning. Same command a bit later gives the correct result:

Wrong Response:

```
2024-09-18 15:35:50.852 DEBUG [p2p]   Handle DATA DATA - Received data {
  stationSN: 'T8170TXXXXXXXX4D',
  seqNo: 2,
  commandIdName: 'CMD_SDINFO_EX',
  commandId: 1144,
  resultCodeName: 'ERROR_PPCS_SUCCESSFUL',
  resultCode: 0,
  resultData: '0000000000000000',                                 <==== Look here!!!
  data: '000000000000000000000000',
  seqNumber: 5,
  p2pDataSeqNumber: 3,
  offsetDataSeqNumber: 1
}
2024-09-18 15:35:50.853 DEBUG [p2p]   Handle DATA DATA - Result data for command received {
  stationSN: 'T8170TXXXXXXXX4D',
  message: {
    sequence: 4,
    commandType: 1144,
    nestedCommandType: undefined,
    channel: 255,
    acknowledged: true,
    retries: 0,
    returnCode: 0,
    data: <Buffer d1 00 00 04 58 5a 59 48 78 04 88 00 00 00 01 00 ff 00 00 00 00 00 00 00 00 00 00 00 63 62 30 37 62 30 65 65 32 63 62 64 64 34 63 34 63 32 30 32 36 37 ... 106 more bytes>,
    customData: undefined
  },
  resultCodeName: 'ERROR_PPCS_SUCCESSFUL',
  resultCode: 0
}
2024-09-18 15:35:50.853 DEBUG [http]  Station got p2p command response {
  station: 'T8170TXXXXXXXX4D',
  commandType: 1144,
  channel: 255,
  returnCodeName: 'ERROR_PPCS_SUCCESSFUL',
  returnCode: 0,
  customData: undefined
}
2024-09-18 15:35:50.854 DEBUG [addon] Event "PropertyChanged": station: T8170TXXXXXXXX4D | name: sdStatus | value: 0
2024-09-18 15:35:50.854 DEBUG [addon] Event "PropertyChanged": station: T8170TXXXXXXXX4D | name: sdCapacity | value: 0
2024-09-18 15:35:50.855 DEBUG [addon] Event "PropertyChanged": station: T8170TXXXXXXXX4D | name: sdCapacityAvailable | value: 0
```

Correct response:

```
2024-09-18 15:40:38.067 DEBUG [p2p]   Handle DATA DATA - Received data {
  stationSN: 'T8170TXXXXXXXX4D',
  seqNo: 3,
  commandIdName: 'CMD_SDINFO_EX',
  commandId: 1144,
  resultCodeName: 'ERROR_PPCS_SUCCESSFUL',
  resultCode: 0,
  resultData: 'e41b0000c81b0000',                                     <==== Look here!!!
  data: '00000000e41b0000c81b0000',
  seqNumber: 6,
  p2pDataSeqNumber: 4,
  offsetDataSeqNumber: 0
}
2024-09-18 15:40:38.067 DEBUG [p2p]   Handle DATA DATA - Result data for command received {
  stationSN: 'T8170TXXXXXXXX4D',
  message: {
    sequence: 5,
    commandType: 1144,
    nestedCommandType: undefined,
    channel: 255,
    acknowledged: false,
    retries: 0,
    returnCode: 0,
    data: <Buffer d1 00 00 05 58 5a 59 48 78 04 88 00 00 00 01 00 ff 00 00 00 00 00 00 00 00 00 00 00 63 62 30 37 62 30 65 65 32 63 62 64 64 34 63 34 63 32 30 32 36 37 ... 106 more bytes>,
    customData: undefined
  },
  resultCodeName: 'ERROR_PPCS_SUCCESSFUL',
  resultCode: 0
}
2024-09-18 15:40:38.068 DEBUG [http]  Station got p2p command response {
  station: 'T8170TXXXXXXXX4D',
  commandType: 1144,
  channel: 255,
  returnCodeName: 'ERROR_PPCS_SUCCESSFUL',
  returnCode: 0,
  customData: undefined
}
2024-09-18 15:40:38.069 DEBUG [addon] Event "PropertyChanged": station: T8170TXXXXXXXX4D | name: sdCapacity | value: 7140
2024-09-18 15:40:38.069 DEBUG [addon] Event "PropertyChanged": station: T8170TXXXXXXXX4D | name: sdCapacityAvailable | value: 7112
```

No idea why :(

**Comment by PhilippEngler** [comment]
Thank you, @bropat. I had some issues with the PKCS1 decryption. Since I have solved it, the problem did not occurred again. I have tied it the last fife day - no problem so far...

Your implementation runs correctly regarding current data and battery consumption. Thank you.
